src/Agent.Worker/StepsRunner.cs

```diff
@@ -72,36 +73,41 @@ public sealed class StepsRunner : AgentService, IStepsRunner
                 step.ExecutionContext.Variables.RecalculateExpanded(out expansionWarnings);
                 expansionWarnings?.ForEach(x => step.ExecutionContext.Warning(x));
 
+                // Record the job cancellation state prior evaluating the condition.
+                bool alreadyCanceled = jobContext.CancellationToken.IsCancellationRequested;
+
                 // Evaluate condition.
                 step.ExecutionContext.Debug($"Evaluating condition for step: '{step.DisplayName}'");
                 var expressionManager = HostContext.GetService<IExpressionManager>();
+                DT.INode conditionTree = null;
                 bool? conditionResult = null;
                 try
                 {
-                    conditionResult = expressionManager.Evaluate(step.ExecutionContext, step.Condition);
+                    conditionTree = expressionManager.Parse(step.ExecutionContext, step.Condition);
```

Comment:
I found a bug while testing and had to move the Parse into the try catch (and treated as critical failure).

Furthermore, when I originally added this PR, I didn't have the validate method. But now that it's there, I think we could validate up front all step conditions for syntax errors, rather than fail in the middle of the build. Then we actually don't need the Parse method inside the try/catch here.

Note, the validate method was added for server side agnostic to agent extensions. If we want to validate up front, then the validate method needs some small tweaks so it doesn't have to be run in "extension-unaware" mode.